### PR TITLE
refactor(NODE-5419): move csfle source and tests into the driver

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

This PR depends on https://github.com/mongodb/libmongocrypt/pull/667.

This is a large PR but there's not much that can be done about that.  I'll highlight the relevant changes here.

- FLE code has been moved from `encryption` into the `src`.  The FLE logic was basically ported directly as-is into `src`, with three modifications
-- we no longer inject `mongodb` into any of the files.  Instead, each file is a top-level module.
-- we no longer inject BSON into any FLE logic.  Each FLE class requires from `bson` directly now.  And the _bson properties have been removed from each class as well.
-- I converted the imports to `import` syntax instead of `require` syntax.  I did this while debugging, it wasn't necessary, but once it was done I figured I'd leave it because it's one fewer thing to adjust when I convert the files to TS (up next).  I can revert this now though, if y'all would like.
- the tests have been ported into `integration/node-specific` and `unit/client-side-encryption`.
-- the unit tests were mostly left as-is, I plan to update them more when they're converted to TS.  I did have to fix imports though.  I also deleted some redundant tests.
-- any tests that required a "live server" have been moved into integration.
-- shared library tests were moved into `integration`, because we have integration testing environments that have the shared library installed.

Some other misc changes

- obviously, the driver has been updated to consume AutoEncrypter directly from `src`.  changes were pretty minimal here
- I moved the FLE import logic into `deps.ts` to align with our existing optional deps pattern.
- the shared library tests now spawn subprocesses, since loading the shared library is a process-wide change and the tests were interfering with each other.
- a new filter was added, GenericPredicateFilter, that abstracts away the need to manually skip tests in before each hooks by running a specified predicate function in the filter.

**NOTE FOR REVIEWERS**

Until an alpha is release of mongodb-client-encryption containing the changes to the FLE bindings, pretty much all CI is red.  The important tests here are the custom dependency tests with a pinned commit, which pass.  

All the integration tests and unit tests pass locally against a local copy the changes in https://github.com/mongodb/libmongocrypt/pull/667 too.  Once we merge the bindings changes, I'll release an alpha, pull the alpha into this PR and rerun CI.

To test the changes:

- make sure you've checked out PR#667 from libmongocrypt.
- make sure your environment is set up for FLE tests (kms servers, credentials, ect)
- build and link the bindings: `FLE_BINDINGS_PATH=<> ./etc/tooling/fle.sh link --build-libmongocrypt`
- run the tests normally

##### Is there new documentation needed for these changes?

Not yet.

### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/tools/unified-spec-runner/unified-utils.ts [hunk 1]
```diff
@@ -1,30 +1,31 @@
+/* eslint-disable @typescript-eslint/no-restricted-imports */
 import { EJSON } from 'bson';
 import { expect } from 'chai';
 import ConnectionString from 'mongodb-connection-string-url';
 import { gte as semverGte, lte as semverLte } from 'semver';
 import { isDeepStrictEqual } from 'util';
 
+import { ClientEncryption } from '../../../src/client-side-encryption/clientEncryption';
```
Comment (nbbeeken) [inline test/tools/unified-spec-runner/unified-utils.ts:8]: Duplicate identifier error here, how's this working?

File: test/types/client-side-encryption.test-d.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+// TODO(NODE-xxxx): add TS support for FLE in the driver and unskip type tests
```
Comment (nbbeeken) [inline test/types/client-side-encryption.test-d.ts:1]: TS ticket TODO

File: test/unit/client-side-encryption/autoEncrypter.test.js [hunk 3]
```diff
@@ -3,42 +3,21 @@
 const fs = require('fs');
 const path = require('path');
 const sinon = require('sinon');
-const mongodb = require('mongodb');
-const BSON = mongodb.BSON;
-const EJSON = BSON.EJSON;
+const { MongoClient } = require('../../../src/mongo_client');
+const bson = require('bson');
+const { EJSON } = bson;
```
Comment (nbbeeken) [inline test/unit/client-side-encryption/autoEncrypter.test.js:8]:
```suggestion
const { MongoClient, BSON } = require('../../../src/mongo_client');
const { EJSON } = BSON;
```

Let's keep obtaining BSON from the driver here

File: test/unit/client-side-encryption/stateMachine.test.js [hunk 4]
```diff
@@ -6,9 +6,10 @@ const tls = require('tls');
 const fs = require('fs');
 const { expect } = require('chai');
 const sinon = require('sinon');
-const mongodb = require('mongodb');
-const BSON = mongodb.BSON;
-const StateMachine = require('../lib/stateMachine')({ mongodb }).StateMachine;
+const { serialize, Long, Int32 } = require('bson');
```
Comment (nbbeeken) [inline test/unit/client-side-encryption/stateMachine.test.js:9]: BSON from driver here

**Review by nbbeeken — CHANGES_REQUESTED** [review]
A few small fix ups, and from slack if we can get the CI green with skips (as long as the cursom version keeps working) we're good to go

File: test/unit/client-side-encryption/autoEncrypter.test.js (hunk 3 above)
Comment (baileympearson) [reply test/unit/client-side-encryption/autoEncrypter.test.js:8]: The mongo client doesn't export BSON.

File: src/client-side-encryption/mongocryptdManager.js [hunk 5]
```diff
@@ -1,12 +1,12 @@
-'use strict';
-
-const spawn = require('child_process').spawn;
+import { spawn } from 'child_process'
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/client-side-encryption/mongocryptdManager.js:1]: Unlike our optional dependencies, child_process (and `once` below) are node APIs.  Addressing these is outside the scope of this work, as well as outside of any existing pattern of usage we have in the driver for optional Node APIs, and so I'm going to leave it as-is without a TODO.

File: test/tools/unified-spec-runner/unified-utils.ts (hunk 1 above)
Comment (baileympearson) [reply test/tools/unified-spec-runner/unified-utils.ts:8]: Our tests aren't compiled

File: test/unit/client-side-encryption/autoEncrypter.test.js (hunk 3 above)
Comment (nbbeeken) [reply test/unit/client-side-encryption/autoEncrypter.test.js:8]: Whoops fixed, the driver export has it

File: src/deps.ts [hunk 6]
```diff
@@ -394,14 +394,21 @@ export interface AutoEncryptionOptions {
...
+type MongoCrypt = { MongoCrypt: any };
+/** A utility function to get the instance of mongodb-client-encryption, if it exists. */
+export function getMongoDBClientEncryption(): MongoCrypt | null {
+  let mongodbClientEncryption = null;
+
+  try {
+    // NOTE(NODE-3199): Ensure you always wrap an optional require literally in the try block
+    // Cannot be moved to helper utility function, bundlers search and replace the actual require call
+    // in a way that makes this line throw at bundle time, not runtime, catching here will make bundling succeed
+    mongodbClientEncryption = require('mongodb-client-encryption');
```
Comment (nbbeeken) [inline src/deps.ts:406]: What is the reason for `MONGODB_CLIENT_ENCRYPTION_OVERRIDE` env variable support being removed here?

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 7]
```diff
@@ -16,6 +17,7 @@ const { coerce, gte } = require('semver');
 const {
   externalSchema
 } = require('../../spec/client-side-encryption/external/external-schema.json');
+const { ClientEncryption } = require('../../../src/client-side-encryption/clientEncryption');
```
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:20]: Exemption is file level, vs the rest being per line

**Review by nbbeeken — CHANGES_REQUESTED** [review]
We've introduced a bunch of linter exemptions for the imports, let's just double check their used consistently eventually someone will h to have to clean them up. 

Q about the FLE import change.

File: src/deps.ts (hunk 6 above)
Comment (nbbeeken) [reply src/deps.ts:406]: from slack: This was for testing b/c there was a circular dependency that cause a BF https://jira.mongodb.org/browse/NODE-4254, there's no longer a driver dep in FLE.

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**